### PR TITLE
Voucher Portal Theme Spec

### PR DESCRIPTION
**Pull request by fservida**
Hello,
I've written a themespec file for handling vouchers from a text voucher roll.
It is very basic, but could be interesting to add it to the main repo so that others have a starting point or just a basic solution for vouchers right on openwrt without needing external servers.

Let me know if you think it is a good idea, and if so what might to be changed; the script is very rough and might benefit from some experienced insight, I've almost no experience with Posix shell scripting.
As I'm passing the voucher input directly to various shell commands, I've tried to validate it to 9 alphanumeric chars and dashes. It seems to work but there might be better ways to validate it.

### CODE REVIEW

**Comment by bluewavenet** [comment]
@fservida 
Thank you very much for this, the effort is much appreciated. Voucher login is a very outdated concept these days but if that is what the customer wants... :-D

This is not really any different from having a preconfigured user database and that is best done using the built in support of the _theme_click-to-continue-custom-placeholders.sh_ or _theme_user-email-login-custom-placeholders.sh_ example scripts, with just a custom validation needing to be added before auth_log() is called.

This is in principle what you have done, but would it be better to base it on the existing built in custom support?

There is one major problem with this PR though. You are writing to the database file every time a client logs in. This is very much to be avoided unless the file is located on some other storage as it will, probably in a few months of use, lead to flash wear failures on the soldered in memory chips found on most router class hardware. Moving to local usb storage is simple enough, but doing that is device dependent of course.

Another less severe problem is "unexpired token" reuse. There is nothing in this PR to update the voucher when a client logs out or has an idle-timeout. Also if a client exceeds a volume quota, there is no means of flagging the voucher. All of this could be done with a binauth script though.

Finally, a quick look at your code shows another, but less scary, bit of text in the footer.

**Comment by fservida** [comment]
I understand the outdated principle, though some times it is sufficient for places like small hotels and bed and breakfasts, where a preprinted list might be available to customers.
I might have misunderstood the comment about the database, but if it relates to the "vouchers.txt" database, unless I missed something in my shell code I should be writing only twice, once upon the first login (when the last column is still 0) and once if the user tries to use an expired voucher, at which point it will be deleted from the file.
For all other login attempts the file should only be read.
If I'm missing something can you point me to the specific lines where you see the behavior?

I agree on the footer, it was also forgotten, although Cthulhu is currently taking over my wifi ^^
Will try to restore the theme looks to match the other themes.

For unexpired tokens, that is handled by the last and before last columns, they specify respectively the time of first use of a voucher and the validity period in minutes. Upon login the first use time (unix timestamp) is added to the validity period of the token and compared to the current time. thus even if expired tokens are present they would not be usable (and are deleted as soon as checked).

I see the problem with quotas though, did not really think it through while implementing it.

**Comment by bluewavenet** [comment]
The number of times the vouchers.txt gets written to is maybe not huge, but it is at least one write per client using the system and occurs **EVERY** time someone logs in. This is extremely bad practice on a flash rom.

These days, with low cost unlimited Internet availability being the norm, the provision of captive portals at public venues is expected and business will be lost if not provided. A simple autonomous portal with client self account creation is the way forward and can simply cover the local legal requirements (even the contradictory requirements in the EU). The overhead of providing customers with a voucher code that is in sync with a voucher database is something non-trivial that businesses are not prepared to accept.
Only in rare niche applications will it be of use, at least in my experience. Is it more common in your location?

This is worth keeping though and I suggest that we create a new folder in the repo where things like this can be submitted for use of others. 
A "Community" folder would be good, similar to the "testing" folder that exists already.
As you said "so that others have a starting point or just a basic solution". 

Perhaps close this PR and resubmit with files in Community/themespec/theme_voucher/ ?

**Comment by fservida** [comment]
It is particularly common for small shops, with limited number of customers, which do not want to go through the whole hassle of setting up a secondary portal for user self onboarding and management, but that still want to retain a small control over who is using the network and when (I've encountered it in small bnb, refuges, campings...).
I agree completely that for bigger venues this is completely unpractical. 

For the flash wear, indeed persistent writing to flash degrades it, however I do still think that for very small shops, it still would not hit the wear out limit; and if one is concerned moving it to external storage might be the best solution.
I've included a README with such a notice and suggestion.

I agree that keeping this in a separate folder might be the best solution, I've committed as needed, let me know it that is satisfying for you.

**Comment by dianariyanto** [comment]
Hi @bluewavenet ,
This might sound weird if someone still uses vouchers to get internet access from hotspot areas (Captive Portal with Voucher Coupon). But in developing countries where the internet is still expensive and the availability of cellular signals is not evenly distributed to every region, hotspot vouchers are still a mainstay for sharing internet subscription fees in my country. This activity is popular in my country as RT/RW NET, a cheap internet movement for villages and remote areas. Currently, the only hardware that supports this is MikroTik which is quite expensive. We hope to have a hotspot voucher server with higher specifications at the same price.

In conclusion, using hotspot vouchers is still very relevant, especially in countries with expensive internet access costs. So, captive portal login is still a good idea. At least in my country for now.

**Comment by bluewavenet** [comment]
@dianariyanto 
I was looking at this PR only yesterday with regards to merging it as a community resource.

> where the internet is still expensive and the availability of cellular signals is not evenly distributed to every region, hotspot vouchers are still a mainstay for sharing internet subscription fees

As I have mentioned before, in the majority of countries, paid access at a business venue has become a guaranteed method of loosing customers. But I can see this is a valid use case where Internet access is very expensive or not generally available, albeit a rapidly shrinking use case.

The problem with this PR is that a database is stored on flash memory. Constant read/write to flash memory is extremely bad practice and will result, sooner or later, in non-repairable damage to the device. 

@fservida has added a warning in the README.md file, but I do not think this is sufficient.
Dynamic data should be stored either on removable media (eg a usb stick) or on /tmp ramdisk with sync to external hosting or some combination of these.

IMO It would be sufficient, for example, if this PR was to have "installation instructions" on how to mount a usb stick and the voucher scripts modified to use that mount point.

**Comment by bluewavenet** [comment]
@fservida 
I have merged this and then modified the script to use the openNDS log location thus preventing a default possibility of flash wear problems.
It is now a very simple matter to override the log location to the mountpoint of some external storage in the standard way used by all ThemeSpec scripts.
The README.md file has been updated accordingly.

**Comment by fservida** [comment]
Awesome, thank you!

**Comment by ganganrobert** [comment]
@fservida Hi! How can I modify this theme (add picture, parameters etc...) ?

**Comment by fservida** [comment]
Hello, you would have to modify the html sections in the different functions in  theme_voucher.sh to reflect the interface changes you need ;)

**Comment by ganganrobert** [comment]
Maybe screenshot for example?)

**Comment by fservida** [comment]
I don't see how a screenshot can be helpful and am unable to provide it anyway as I'm not running this system anymore, but just check the voucher_form() function in the mentioned file, you have the HTML for rendering the voucher input section, or the header() or footer() sections.
